### PR TITLE
[c10d] Move sync_params to C++

### PR DESCRIPTION
**Pull request by goldsborough**
The next function I'm moving to C++ is `sync_params`. It is stacked on top of https://github.com/pytorch/pytorch/pull/9729, so some changes will go away when it lands and I rebase.

I also split code into a `.h` and `.cpp` file for better code organization.

@teng-li @pietern @apaszke

### CODE REVIEW

**Comment by weiyangfb** [comment]
cc @teng-li would you like to review this PR too?

**Comment by teng-li** [comment]
@goldsborough We do have a test covering this change right?

**Comment by goldsborough** [comment]
@teng-li I ran the c10d DDP test that I wrote. Should we also write some unit tests for these individual functions?

File: test/test_c10d.py [hunk 1]
```diff
@@ -509,10 +515,103 @@ def test_gloo_backend(self):
...
+    @skip_if_not_multigpu
+    def test_sync_params_no_buffers(self):
+        # Set up process group.
+        store = c10d.TCPStore('localhost', self.port, self.is_master)
+        options = c10d.ProcessGroupGloo.Options()
+        options.devices = [c10d.ProcessGroupGloo.create_tcp_device(interface="lo")]
+        process_group = c10d.ProcessGroupGloo(store, self.rank, self.world_size, options)
+
+        # Use all available devices on every process here (data is small, so should be fine).
+        devices = list(range(torch.cuda.device_count()))
```
Comment (teng-li) [inline test/test_c10d.py]: I am thinking of dividing the devices among all processes, wdyt?

File: test/test_c10d.py [hunk 2]
```diff
@@ -509,10 +515,103 @@ def test_gloo_backend(self):
...
+            buffer_data += [torch.zeros(10, device=torch.device('cuda', d)).chunk(2) for d in devices[1:]]
+        else:
+            buffer_data = [torch.zeros(10, device=torch.device('cuda', d)).chunk(2) for d in devices]
+
+        c10d._sync_params(
+            process_group,
+            parameter_data=parameter_data,
+            buffer_data=buffer_data,
+            devices=devices,
+            broadcast_bucket_size=25 * MB,
```
Comment (teng-li) [inline test/test_c10d.py]: will 25MB be too big in this case? Ideally, we would like to have more than 1 bucket to be created.

**Review by teng-li — APPROVED** [review]
Look great, thank you for adding those test cases.

Please see my comments.

File: test/test_c10d.py (hunk 1 above)
Comment (goldsborough) [reply test/test_c10d.py]: We do this in other tests, but in a lot of cases we only run with two GPUs, so with two processes that would only leave one device per process and make this test less meaningful. Is this reasonable?

File: test/test_c10d.py (hunk 2 above)
Comment (goldsborough) [reply test/test_c10d.py]: Good point

File: test/test_c10d.py (hunk 1 above)
Comment (teng-li) [reply test/test_c10d.py]: it should be fine for Gloo, but NCCL it won't work AFAIK. We can leave as is

**Comment by goldsborough** [comment]
@apaszke @ezyang @soumith looking for codeowner approval

**Review by soumith — APPROVED** [review]

**Comment by soumith** [comment]
might possibly need a rebase. cc: @teng-li

**Pull request closed without merging**